daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/Convert.hs

```diff
@@ -353,9 +361,30 @@ convertChoice env (VarIs "C:Choice" `App` Type tmpl `App` Type (TypeCon chc [])
             ETmLam thisBndr (ETmLam argBndr body)
               | fst argBndr `Set.notMember` cata freeVarsStep body ->
                 ETmLam thisBndr body `ETmApp` thisVar
-            _ ->
-                controllerExpr `ETmApp` thisVar `ETmApp` argVar
-    chcUpdate <- fmap (\u -> u `ETmApp` thisVar `ETmApp` selfVar `ETmApp` argVar) (convertExpr env choice)
+            _ -> controllerExpr `ETmApp` thisVar `ETmApp` argVar
+    expr <- fmap (\u -> u `ETmApp` thisVar `ETmApp` selfVar `ETmApp` argVar) (convertExpr env choice)
+    chcUpdate <-
+      if not $ postConsuming
+        then
+          pure expr
+      else
+        do
+           -- NOTE(SF): Support for 'postconsuming' choices. The idea
+           -- is to evaluate the user provided choice body and
+           -- following that, archive. That is, in pseduo-code, we are
+           -- going for an expression like this:
+           --     expr this self arg >>= \res ->
+           --     archive signatories self >>= \_ -> return res
+          let archive =
+                EUpdate $
+                  UExercise tmplTyCon'
+                            (mkChoiceName "Archive")
+                            selfVar signatories mkEUnit
+              returnRes = EUpdate $ UPure chcReturnType (EVar $ mkVar "res")
+          pure $
+            EUpdate $
+              UBind (Binding (mkVar "res", chcReturnType) expr)
+                    (EUpdate $ UBind (Binding (mkVar "_", TUnit) archive) returnRes)
```

Comment:
I find the way you've broken this up in the code a bit odd. I tried doing code golf and got to something more like:

```
let callArchive = EUpdate $ UExercise tmplTyCon' (mkChoiceName "Archive") selfVar signatories mkEUnit
EUpdate $ UBind (Binding (mkVar "res", chcReturnType) expr) $
EUpdate $ UBind (Binding (mkVar "_", TUnit) callArchive) $
EUpdate $ UPure chcReturnType (EVar $ mkVar "res")
```

Reply:
✔️